### PR TITLE
Nmark tactic

### PR DESCRIPTION
**Pull request by kfu02**
## Description
This implements the nmark tactic, which marks the n enemies closest to the ball with the robots of ours that are closest to them. For now, ignore the roughness of the `defensive_clear.py` file, it will be uncommented and filled in once pivot_kick from #1644  is merged in.

## Associated Issue
https://app.clickup.com/t/y3fvx9

### Test Case 1
1. Start sim
2. Drag enemy robots (blue) around the field
3. Kick ball around

Expected result: Our 2 closest robots to the 2 enemies closest to the ball should move to mark those enemies, regardless of where the ball is. (I know this wording is a little unclear but the behavior makes sense when you see it.)

### Test Case 2
1. Change the number in the `init()` of `play/defensive_clear.py` that controls how many robots nmark handles.
2. Repeat Test Case 1.

Expected result: NMark works the same, but with more (or less) than 2 robots.

### CODE REVIEW

File: rj_gameplay/rj_gameplay/play/defensive_clear.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    	flat_results = self.role_assigner.assign_roles(flat_requests, world_state, prev_results)
+    	role_results = play.unflatten_results(flat_results)
+
+    	# Get list of all skills with assigned roles from tactics
+
+    	# skills = self.striker_tactic.tick(role_results[self.striker_tactic]) + self.two_mark.tick(role_results[self.two_mark])
+    	skills = self.two_mark.tick(role_results[self.two_mark])
+    	skill_dict = {}
+    	# skill_dict.update(role_results[self.striker_tactic])
+    	skill_dict.update(role_results[self.two_mark])
```
Comment (kylestach) [inline rj_gameplay/rj_gameplay/play/defensive_clear.py:45]: I'd appreciate it if you could remove the dead code.

File: rj_gameplay/rj_gameplay/play/defensive_clear.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        self,
+        world_state: rc.WorldState,
+        prev_results: role.assignment.FlatRoleResults,
+        props,
+    ) -> Tuple[Dict[Type[tactic.SkillEntry], List[role.RoleRequest]], List[tactic.SkillEntry]]:
+
+    	# Get role requests from all tactics and put them into a dictionary
+    	role_requests: play.RoleRequests = {}
+    	# role_requests[self.striker_tactic] = self.striker_tactic.get_requests(world_state, None)
+    	role_requests[self.two_mark] = (self.two_mark.get_requests(world_state, None))
```
Comment (kylestach) [inline rj_gameplay/rj_gameplay/play/defensive_clear.py:32]: Shouldn't this have a tactic for the robot kicking the ball? And also in theory a pass seeking tactic.

File: rj_gameplay/rj_gameplay/tactic/nmark_tactic.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        ]
+        
+    def compute_props(self):
+        pass
+
+    def create_request(self, **kwargs) -> role.RoleRequest:
+        """Creates a sane default RoleRequest.
+        :return: A list of size 1 of a sane default RoleRequest.
+        """
+        pass
```
Comment (kylestach) [inline rj_gameplay/rj_gameplay/tactic/nmark_tactic.py:73]:
```suggestion
```

**Review by kylestach — CHANGES_REQUESTED** [review]
Mostly good, a few minor comments.

File: rj_gameplay/rj_gameplay/play/defensive_clear.py (hunk 2 above)
Comment (kfu02) [reply rj_gameplay/rj_gameplay/play/defensive_clear.py:32]: Yes and yes. I started off writing the defensive clear play only to realize both the nmark and striker tactics weren't written yet. So this PR is step 1. For step 2 I'll write a striker tactic, uncomment the aforementioned "dead code", and possibly add the pass seeker, then put up a new PR for the full defensive clear play.

File: rj_gameplay/rj_gameplay/tactic/nmark_tactic.py (hunk 3 above)
Comment (kfu02) [reply rj_gameplay/rj_gameplay/tactic/nmark_tactic.py:73]: should be fixed at the interface level, as discussed on Teams, ignoring for now

Comment (kfu02) [reply rj_gameplay/rj_gameplay/tactic/nmark_tactic.py:73]: issue #1665 made

File: rj_gameplay/rj_gameplay/tactic/nmark_tactic.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
+from dataclasses import dataclass
+from typing import List, Optional
+
+import stp.action as action
```
Comment (Yudai-8-om) [inline rj_gameplay/rj_gameplay/tactic/nmark_tactic.py]: I believe action is imported but not used

**Review by Yudai-8-om — APPROVED** [review]
LGTM! just added one comment.

File: rj_gameplay/rj_gameplay/tactic/nmark_tactic.py (hunk 4 above)
Comment (kfu02) [reply rj_gameplay/rj_gameplay/tactic/nmark_tactic.py]: It is not, nice catch

Comment (kfu02) [inline rj_gameplay/rj_gameplay/tactic/nmark_tactic.py]:
```suggestion
```

File: rj_gameplay/rj_gameplay/tactic/nmark_tactic.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        self,
+        robot: rc.Robot,
+        prev_result: Optional["RoleResult"],
+        world_state: rc.WorldState,
+    ) -> float:
+
+        # TODO: prevent gameplay crashing w/out this check
+        if robot is None or self.enemy_to_mark is None: 
+            return 0
+        return np.linalg.norm(robot.pose[0:2]-self.enemy_to_mark.pose[0:2])
```
Comment (kfu02) [inline rj_gameplay/rj_gameplay/tactic/nmark_tactic.py]: TODO: cost should be in sec, divide by `global_parameters.soccer.robot.max_speed`

**Review by kfu02 — COMMENTED** [review]
cost nit

**Review by kylestach — APPROVED** [review]
LGTM

**Pull request merged**